File: lib/core/transformers/standard/transformer_block.dart
/// Decoder-only GPT-style transformer block (pre-norm).
///
///     x_norm1 = LN1(x)
///     attn    = MHA(x_norm1)
///     x_res1  = x + attn
///     x_norm2 = LN2(x_res1)
///     ffn     = FFN(x_norm2)
///     out     = x_res1 + ffn
import 'package:dart_cuda/core/attention/multi_head_attention.dart';
import 'package:dart_cuda/core/layers/feed_forward.dart';
import 'package:dart_cuda/core/layers/layer_norm.dart';
import 'package:dart_cuda/core/layers/nn.dart';
import 'package:dart_cuda/core/tensor/gpu_tensor.dart';

class TransformerBlock extends Module {
  final int embedSize;
  final MultiHeadAttention attention;
  final FeedForward ffn;
  final LayerNorm ln1;
  final LayerNorm ln2;

  TransformerBlock(this.embedSize, int numHeads, {bool masked = false})
      : attention = MultiHeadAttention(numHeads, embedSize, masked: masked),
        ffn = FeedForward(embedSize),
        ln1 = LayerNorm(embedSize),
        ln2 = LayerNorm(embedSize);

  Tensor forward(Tensor x, List<Tensor> tracker) {
    final xNorm1 = ln1.forward(x, tracker);
    final attnOut = attention.forward(xNorm1, tracker);
    final xRes1 = x + attnOut;
    tracker.addAll([attnOut, xRes1]);

    final xNorm2 = ln2.forward(xRes1, tracker);
    final ffnOut = ffn.forward(xNorm2, tracker);
    final out = xRes1 + ffnOut;
    tracker.add(ffnOut);
    return out;
  }

  @override
  List<Tensor> parameters() => [
        ...attention.parameters(),
        ...ffn.parameters(),
        ...ln1.parameters(),
        ...ln2.parameters(),
      ];
}

File: python/dart_cuda/core/transformers/standard/transformer_block.py
"""Decoder-only GPT-style transformer block.

Port of `lib/transformer/transformer_block.dart` (dart-torch).
Pre-norm: LN -> MHA -> residual -> LN -> FFN -> residual.
"""

from __future__ import annotations

from typing import List

from ...attention.multi_head_attention import MultiHeadAttention
from ...layers.feed_forward import FeedForward
from ...layers.layer_norm import LayerNorm
from ...layers.nn import Module
from ...tensor.gpu_tensor import Tensor


class TransformerBlock(Module):
    def __init__(self, embed_size: int, num_heads: int, *, masked: bool = False):
        self.embed_size = embed_size
        self.attention = MultiHeadAttention(num_heads, embed_size, masked=masked)
        self.ffn = FeedForward(embed_size)
        self.ln1 = LayerNorm(embed_size)
        self.ln2 = LayerNorm(embed_size)

    def forward(self, x: Tensor, tracker: List[Tensor]) -> Tensor:
        x_norm1 = self.ln1.forward(x, tracker)
        attn_out = self.attention.forward(x_norm1, tracker)
        x_res1 = x + attn_out
        tracker.extend([attn_out, x_res1])

        x_norm2 = self.ln2.forward(x_res1, tracker)
        ffn_out = self.ffn.forward(x_norm2, tracker)
        out = x_res1 + ffn_out
        tracker.extend([ffn_out, out])
        return out

    def parameters(self) -> List[Tensor]:
        return [
            *self.attention.parameters(),
            *self.ffn.parameters(),
            *self.ln1.parameters(),
            *self.ln2.parameters(),
        ]
